stage #2.5/module #2 ORM - JPA, Hibernate/#6. Hibernate Caching/content.md

```diff
@@ -0,0 +1,109 @@
+* in clustered environments.
+* **Transactional** - It prevents stale data, so should be used when stale data is not acceptable at all.
+
+ </br>
+  It needs to tell the hibernate itself what exactly to cache. Luckily, this can be done very easily with
+annotations, like so:
+    
+    @Entity
+    @Table(name = "customer")
+    @Cache(usage = CacheConcurrencyStrategy.READ_WRITE)
+    public class Customer {
+    @Id
+    @GeneratedValue
+    private Long id;
+    private String name;
+    @OneToMany(mappedBy = "order")
+    private List<Order> orders = new ArrayList<>();
+    ...
+    }
+The class dependencies aren't catch by default. For example, if we consider the class above - Customer, then when 
+fetching, the orders' collection will be taken from the database, and not from the second-level cache. If you want to
+cache dependencies as well, then the class should look like this:
+
+    @Entity
+    @Table(name = "customer")
+    @Cache(usage = CacheConcurrencyStrategy.READ_WRITE)
+    public class Customer {
+    @Id
+    @GeneratedValue
+    private Long id;
+    private String name;
+    @OneToMany(mappedBy = "order")
+    @Cache(usage = CacheConcurrencyStrategy.READ_WRITE)
+    private List<Order> orders = new ArrayList<>();
+    ...
+    }
+**How it works**: When hibernate session try to load an entity, it will first find into the first-level cache,
+if it does not found then it will look into the second-level cache and return the response (if available), but before
+returning the response it will store that object/data into first-level also so next time no need to come at the 
+session-level. When data is not found in the second level then it will go to the database to fetch data. Before 
+returning a response to the user it will store that object/data into both levels of cache so next time it will be 
+available at cache stages only.
+### 6.4. Query Level Cache
+By default the data of  HQL queries are not cached. If application fires HQL statement in the same or different session,
+multiple times, then data of the query can be cached.
+
+To do so we need to enable, we need to first set **_hibernate.cache.use_query_cache="true"_** in **_hibernate.cfg.xml file**_ .
+Setting up this property will create additional two required cache regions. One to store the data and one to hold the 
+last updated timestamps.
+    
+    <property name="hibernate.cache.use_query_cache" value="true"/>
+
+Once done, we need to explicitly call **_setCacheable(true)_** on Query object before execution:
+
+    Query query = session.createQuery("select c from Customer c where c.name = ?");
+    query.setCacheable(true);
+
+The query cache is similar to the second level cache. But unlike it, the key to the cache data is not the object 
+identifier, but the set of query parameters. And the data itself is the identifiers of the objects that match the query 
+criteria. Thus, it is rational to use this cache with the second-level cache.
```

Comment:
Why don't you use picture in media folder?

Reply:
corrected